### ISSUE TITLE
Getting unexpected result when using element.all(by.xpath()) to read table data

### ISSUE THREAD

**Issue by akbbhatt** [angular/protractor#5447]
My application contains two mat-tables on it. 
I would like to read the table data from my second table

await this.secondTableElement.all(by.xpath(`//tr[${rowIndex}]/td[${columnIndex}]`));

but this statement actually provides me the data from the first table and not from the secondtableElement. I believe when you execute all() method under a parent element, protractor is supposed to search under this element and not from the very beginning of the HTML document.

**Comment by Fuun347** [comment]
await elem.all will return an ElementFinder[]. Prefered way to do this in protractor would be to use ElementArrayFinder. For that remove the await. To get an ElementFinder at index 1 you will need to type: `this.secondTableElement.all(by.xpath(//tr[${rowIndex}]/td[${columnIndex}])).get(1);` This will return an ElementFinder for the second table.

**Comment by akbbhatt** [comment]
Thanks for the reply Funn347. Sorry for providing such a bad example. 
So you mean to tell that if I do something like this also:

this.secondTableElement.all(by.xpath(//tr[${rowIndex}]]))

then I will get "tr"s from all the tables from the page? I was with the impression that protractor will search for tr's only under this secondTableElement.

**Comment by Fuun347** [comment]
Well it will only search for them inside of the parent, yes. I deleted the comment as it was not exactly correct. Are you selecting the table correctly? Does this.secondTableElement reference the correct table?

**Comment by akbbhatt** [comment]
Yes Funn347, I double-checked, in both the cases the this.secondTableElement is referring to the second occurring table only.

**Comment by Fuun347** [comment]
Then there is no reason why this should be selecting the wrong table. Have tried using by.css rather than by.xpath?

**Comment by akbbhatt** [comment]
Yes, it works using by.css but not working with by.xpath.

**Comment by Fuun347** [comment]
Xpath is not recommended at all for any use by the protractor team. This is a section from the protractor style guide
![image](https://user-images.githubusercontent.com/28909459/83875434-bc2c0080-a73f-11ea-9019-0e5a84ca17e0.png)
which you can locate here https://www.protractortest.org/#/style-guide.
Generally I've had issues with xpath before so I absolutely never use it. I'm not sure if this here is a bug or just xpath's broken behavior, but if you can go by without using it, then do go that route.
